### ISSUE TITLE
Tune hyperparameters of learning_rate, batch_size, num_training_steps and nproc-per-node to set up a fair comparison between experiments

### ISSUE THREAD

**Issue by hlnull** [swookey-thinky/xdiffusion#5]
Hi, I'm currently tuning the hyperparameters and am looking for guidance on how to set up a fair comparison between experiments.

Specifically, I'm wondering if it's appropriate to compare two experiments based on the same product of `batch_size*nproc-per-node* num_training_steps`. 

While this approach ensures consistency in terms of overall training volume, I suspect it might not be the whole picture.
For instance, should we consider using different learning rates for these experiments? If so, what are the principles behind setting the learning rate? Any insights would be greatly appreciated.

**Comment by hlnull** [comment]
I noticed the following comments in training/video/autoencoder.py. 

```
    # Now create the optimizer. The optimizer choice and parameters come from
    # the paper:
    # "We tried Adam [31] and RMSProp early on in our experimentation process and chose the
    #  former. We left the hyperparameters to their standard values. We set the learning
    #  rate to 2 × 10−4 without any sweeping, and we lowered it to 2 × 10−5
    #  for the 256 × 256 images, which seemed unstable to train with the larger learning rate."
    optimizers = vae.configure_optimizers(learning_rate=4.5e-6)
```

I’m a bit puzzled by the discrepancy between the suggested learning rate of 2×10⁻⁴ in the comments and the actual value of 4.5×10⁻⁶ used in the code.

Is there a specific reason why 4.5×10⁻⁶ might be considered a better choice? Or is it simply a matter of context or experimental results? I’d appreciate any clarification on this.

**Comment by swookey-thinky** [comment]
Ah good catch here! That comment, and the quote, is pretty outdated, and comes specifically from the original DDPM paper (https://arxiv.org/abs/2006.11239), and has persisted throughout the training code in both video, audio, image, and autoencoder training scripts. I copy/pasted the original training script I wrote, and modified it over time, and left that comment in out of laziness. I don't recall specifically where that value if 4.5e-6 comes from for the autoencoder, but I also used it for the autoencoder in the Make-An-Audio implementation, so likely I just copied it from there. In other words, the comment is a cut/paste relic and the actual value worked somewhere else and i just copied it here, saw it worked as well, and left it in. Ideally I'll move this into the configuration files, but haven't gotten around to it yet.

You bring up a good point though, which is I have done zero hyperparameter tuning for any of the implementations in here. Where possible, I just use the hyperparameters from the papers, and if not, I just use something that "works good enough". Regarding your first comment above, I don't think comparing `batch_size*nproc-per-node* num_training_steps` directly holds that much weight, other than like you mentioned measuring the training throughput. In particular, `num_training_steps` can change pretty dramatically depending on the model architecture - empirically UNet based models seem to converge much faster than the transformer based models, but this doesn't seem to have any effect on overall model quality. Another big missing piece is there is no quantifiable way to know when training has converged in the current codebase. Typically, we should be measuring validation loss after every N steps, and train until the validation loss stops improving. With that measurement in place, we could sweep through the existing parameters until validation loss convergence, and find the best combination that yields the lowest validation loss. However, after a certain point, it's unclear how much improvement in validation loss leads to better image/video quality, so again its not a straight forward comparison. Negative log-likelihood on the validation set (or a set of generated exampled) is typically a better approximation for overall distribution matching quality, so that might yield a better hyperparameter tuning metric, or at least would be a good first pass.

However, an even bigger missing piece is I have not added any aesthetic score or visual quality evaluation metrics - how good are the images that are generated? This is something I have wanted to add for awhile, so that we can better compare the different models. Typically, this would be FID on images and something more comprehensive like VBench (https://github.com/Vchitect/VBench) for video. The challenge is that our datasets here (MNist, Moving MNist) don't really lend themselves too well to the traditional aesthetic quality evaluations. So still thinking about a better approach here.

Open to any suggestions or ideas you have!

**Comment by geronimi73** [comment]
> However, an even bigger missing piece is I have not added any aesthetic score or visual quality evaluation metrics

One quick and easy way would be clipscore for images, with `torchmetrics.functional.multimodal.clip_score`

Hello everyone btw 👋

**Comment by hlnull** [comment]
> Ah good catch here! That comment, and the quote, is pretty outdated, and comes specifically from the original DDPM paper (https://arxiv.org/abs/2006.11239), and has persisted throughout the training code in both video, audio, image, and autoencoder training scripts. I copy/pasted the original training script I wrote, and modified it over time, and left that comment in out of laziness. I don't recall specifically where that value if 4.5e-6 comes from for the autoencoder, but I also used it for the autoencoder in the Make-An-Audio implementation, so likely I just copied it from there. In other words, the comment is a cut/paste relic and the actual value worked somewhere else and i just copied it here, saw it worked as well, and left it in. Ideally I'll move this into the configuration files, but haven't gotten around to it yet.
> 
> You bring up a good point though, which is I have done zero hyperparameter tuning for any of the implementations in here. Where possible, I just use the hyperparameters from the papers, and if not, I just use something that "works good enough". Regarding your first comment above, I don't think comparing `batch_size*nproc-per-node* num_training_steps` directly holds that much weight, other than like you mentioned measuring the training throughput. In particular, `num_training_steps` can change pretty dramatically depending on the model architecture - empirically UNet based models seem to converge much faster than the transformer based models, but this doesn't seem to have any effect on overall model quality. Another big missing piece is there is no quantifiable way to know when training has converged in the current codebase. Typically, we should be measuring validation loss after every N steps, and train until the validation loss stops improving. With that measurement in place, we could sweep through the existing parameters until validation loss convergence, and find the best combination that yields the lowest validation loss. However, after a certain point, it's unclear how much improvement in validation loss leads to better image/video quality, so again its not a straight forward comparison. Negative log-likelihood on the validation set (or a set of generated exampled) is typically a better approximation for overall distribution matching quality, so that might yield a better hyperparameter tuning metric, or at least would be a good first pass.
> 
> However, an even bigger missing piece is I have not added any aesthetic score or visual quality evaluation metrics - how good are the images that are generated? This is something I have wanted to add for awhile, so that we can better compare the different models. Typically, this would be FID on images and something more comprehensive like VBench (https://github.com/Vchitect/VBench) for video. The challenge is that our datasets here (MNist, Moving MNist) don't really lend themselves too well to the traditional aesthetic quality evaluations. So still thinking about a better approach here.
> 
> Open to any suggestions or ideas you have!

Impressive and solid thinking. 

>  Typically, we should be measuring validation loss after every N steps, and train until the validation loss stops improving. With that measurement in place, we could sweep through the existing parameters until validation loss convergence, and find the best combination that yields the lowest validation loss. 

Yes, that’s exactly what we did in the tasks of object detection, image classification, and image segmentation back in the old days. 😄

Will read the resource (such as VBench) in your post and share the results if I come up with something conclusive.

**Comment by swookey-thinky** [comment]
> > However, an even bigger missing piece is I have not added any aesthetic score or visual quality evaluation metrics
> 
> One quick and easy way would be clipscore for images, with `torchmetrics.functional.multimodal.clip_score`
> 
> Hello everyone btw 👋

Hello! Thanks for taking a look :) CLIP score is a great metric for semantic alignment with the text caption. Its can also work as a reconstruction quality metric for image autoencoders, since its tends to focus more on the high level semantic details of the image and less so on low level metrics like L1 or L2 loss at the pixel level, which can be distracted by low frequency artifacts. I like where you are going with this though, adding basic support for torchmetrics based quality evaluations would be fun.

**Comment by swookey-thinky** [comment]
> > Typically, we should be measuring validation loss after every N steps, and train until the validation loss stops improving. With that measurement in place, we could sweep through the existing parameters until validation loss convergence, and find the best combination that yields the lowest validation loss.
> 
> Yes, that’s exactly what we did in the tasks of object detection, image classification, and image segmentation back in the old days. 😄
> 

Exactly! All the old stuff becomes new again :) The main issue with ablating hyperparameters is just the cost involved, in the size of the models, the size of the datasets, and the time it takes to train. In classification, there was solid evidence that we could ablate the parameters on smaller versions of the model, and these would generally hold as we scaled up (or we could derive specific rules, such as batch size scaling, that we could apply). I just haven't seen this as much with diffusion modeling. The first rigorous analysis was the EDM paper (https://arxiv.org/abs/2206.00364) for image diffusion. I just haven't seen much since, and especially not for video diffusion models. Maybe Neural Architecture Search will come back into fashion ;)

With that said, one of the reasons why I wanted to make these smaller implementations on smaller datasets was specifically so that we could explore and test ideas quicker and cheaper. I can afford to train on a T4 or an A10 for a couple dollars, but I want to still be able to understand, run, test, explore, and improve on the state of the art models, which are all built for much larger datasets and GPUs.
